iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessor.java

```diff
@@ -256,6 +276,24 @@ public TDataNodeRegisterResp registerDataNode(TDataNodeRegisterReq req) {
 
   @Override
   public TDataNodeRestartResp restartDataNode(TDataNodeRestartReq req) {
+    // If the remote server is configured to bind to all ip addresses (0.0.0.0),
+    // update the request to use the remote address we got this request from.
+    if ((req.dataNodeConfiguration.getLocation() != null)
```

Comment:
In the current version, we allow an RPCAddress of 0.0.0.0 to join the cluster; for example, the default RPCAddress for 1c1d is 0.0.0.0.
Here we don't have to do this replacement, we can support this node to join, because we can use internalAddress to remove the node without any problems

Reply:
Well ... the issue I was trying to fix, was that in the case of 3 dataNodes joining the cluster, they all will be listed as "0.0.0.0" ... if you then want to remove it by it's real ip then you will not be able to remove it as it only finds 0.0.0.0 nodes and if you use 0.0.0.0, you will delete all nodes. So I think for a 1C1D it should be 127.0.0.1 and not 0.0.0.0 because one is a real ip and the other one isn't.

Reply:
We can use internalAddress to remove it,and there internalAddress must be different

Reply:
And what happens to the second and third data-node that registers? I mean ... two of the 3 issues mentioned in this issue are bound to the fact that 0.0.0.0 is simply not right. I also don't quite understand what's the benefit of using an address that brings no additional information. If we have it the way I implemented it, a user can query the list of data-nodes and instantly knows IP addresses he can connect to. 

A perfect solution would be to list all known public addresses of a data-node, when using 0.0.0.0 in the configuration. This would especially help, if the cluster had a separate network and clients are not part of that. 

I personally think using 0.0.0.0 is the worst possible option.

Reply:
The 0.0.0.0 rpc address (which I think "external address" is a better name) is actually not useful to ConfigNode at all, so theoretically it's ok to replace it to actual ip.

But, this modification should be further considered. What you are trying to do is to correct the poor network configuration of DataNode on ConfigNode, which might cause other issues. 

For example, can ConfigNode ensure that the RequestContext parsed is always consistent? If the RequestContext is inconsistent after two restarts, ConfigNode might perceive it as not the same DataNode. Besides, there is also a need to consider the compatibility with past DataNodes that have joined the cluster via ip 0.0.0.0, and whether ports outside of the external port also need to be corrected. 

We can discuss later whether it is necessary to correct all these addresses, and if it is, this won't be a minor modification. Perhaps it would be better for this PR to limit the scope of changes to fixing the remove DataNode function itself ? That will ensure the atomicity of the PR :)

Reply:
If you ask me, I think handling 2 bits of information instead of only 1 would be the better solution.
Currently a user runs "show datanodes" he possibly gets a list of "0.0.0.0" which helps him nothing. From a user perspective I think a list of public-facing ips would be ideal (So returing the list of real IPs that a user can connect to the RPC server to) ... but on the other side we also have the internal network and as far as I understood it, the Config Node(s) use the internal rpc to do the cluster management (These can be on completely isolated newtorks, that a user can't even reach). So having a list of external rpc addresses per node and one internal rpc address seems the best option.

Reply:
Yeah ... because I think both are important bits of information.

Reply:
How about adding two columns to show datanodes this time, and leaving the 0.0.0.0 replacement as TODO?

Reply:
Part of the whole problem this issue wanted to address, was the fact that 0.0.0.0 was used. I think I don't quite understand why it would be good to keep the 0.0.0.0, as it has the least bit of information.

Reply:
Hi, Each datanode has an external RPC and an internal RPC. The external RPC is meant to be easy to use, so it's possible to use 0.0.0.0 so that it doesn't have to worry about specific IP addresses, which makes it easier for users to use. However, 0.0.0.0 cannot be used for internal IP, so it can be used to identify nodes。

What's your opinion?

Reply:
As I just wrote in the other place. I would allow 0.0.0.0 for setting up the data-node ... but I would use the technique that I use in this pr for using the internal RPC based on the IP used to connect to the config node and I would resolve the 0.0.0.0 on the external rpc to the list of real IP addresses ... you can't open a connection to 0.0.0.0 from any client (possibly you can only do that on the same node as the data-node is running on)